iris-mpc-cpu/src/network/grpc.rs

```diff
@@ -29,94 +28,309 @@ use tonic::{
-        &self,
-        session_id: SessionId,
-        receiver_id: Identity,
-        stream: Sender,
-    ) {
-        self.streams
-            .write()
-            .await
-            .insert((session_id, receiver_id), Arc::new(stream));
+    fn add_session_stream(&mut self, session_id: SessionId, receiver_id: Identity, stream: Sender) {
+        self.streams.insert((session_id, receiver_id), stream);
     }
 
-    async fn get_stream(
-        &self,
-        session_id: SessionId,
-        receiver_id: Identity,
-    ) -> eyre::Result<Arc<Sender>> {
+    fn get_stream(&self, session_id: SessionId, receiver_id: Identity) -> eyre::Result<Sender> {
         self.streams
-            .read()
-            .await
             .get(&(session_id, receiver_id.clone()))
             .ok_or(eyre!(
                 "Streams for session {session_id:?} and receiver {receiver_id:?} not found"
             ))
-            .map(Arc::clone)
+            .cloned()
     }
 
-    async fn count_receivers(&self, session_id: SessionId) -> usize {
+    fn count_receivers(&self, session_id: SessionId) -> usize {
         self.streams
-            .read()
-            .await
             .iter()
             .filter(|((sid, _), _)| *sid == session_id)
             .count()
     }
 }
 
+struct SendTask {
+    value: Vec<u8>,
+    receiver: Identity,
+    session_id: SessionId,
+}
+
+struct ReceiveTask {
+    sender: Identity,
+    session_id: SessionId,
+}
+
+struct StartMessageStreamTask {
+    sender: Identity,
+    session_id: SessionId,
+    stream: UnboundedReceiver<SendRequest>,
+}
+
+struct ConnectToPartyTask {
+    party_id: Identity,
+    address: String,
+}
+
+enum GrpcTask {
+    ConnectToParty(ConnectToPartyTask),
+    CreateSession(SessionId),
+    StartMessageStream(StartMessageStreamTask),
+    IsSessionReady(SessionId),
+    Send(SendTask),
+    Receive(ReceiveTask),
+}
+
+enum MessageResult {
+    Empty,
+    IsSessionReady(bool),
+    ReceivedBytes(Vec<u8>),
+}
+
+struct MessageJob {
+    task: GrpcTask,
+    return_channel: oneshot::Sender<eyre::Result<MessageResult>>,
+}
+
+// Concurrency handler for networking operations
+#[derive(Clone)]
+pub struct GrpcHandle {
+    job_queue: mpsc::Sender<MessageJob>,
+    party_id: Identity,
+}
+
+impl GrpcHandle {
+    pub async fn new(mut grpc: GrpcNetworking) -> eyre::Result<Self> {
+        let party_id = grpc.party_id.clone();
+        let (tx, mut rx) = tokio::sync::mpsc::channel::<MessageJob>(1);
+
+        // Loop to handle incoming tasks from job queue
+        tokio::spawn(async move {
+            while let Some(job) = rx.recv().await {
+                let job_result = match job.task {
+                    GrpcTask::Send(task) => grpc
+                        .send(task.value, &task.receiver, &task.session_id)
+                        .map(|_| MessageResult::Empty),
+                    GrpcTask::Receive(task) => grpc
+                        .receive(&task.sender, &task.session_id)
```

Comment:
This will block everything until the message is received. Effectively that makes the whole process sequential. For example the sends of other sessions will have to wait in line.

Instead, this must spawn a new task and not await. The `return_channel` goes into the task. Then, in a long, long time (for a CPU), the message will show up, and the background task will move it to the return channel.

There are other approaches possible too but that is the simplest right here.

Reply:
This does not apply to Send because it is not async.

This does applies to the other operations too, although they are less performance critical, but you could do this same spawn approach for all operations (except Send).

Reply:
Yep, I see the problem now...

The problem is that spawning a new thread will cause related parts of `GrpcNetworking` to be wrapped up in a thread-safe structs, which will cause locking. So to reduce it to the minimum, I'd go with spawning only `receive` as it'll cause wrapping only `UnboundedReceiver` of `MessageQueueStore` into `Mutex`.

Reply:
Done, see the most recent commit.